cmd/snap-confine/ns-support.c

```diff
@@ -767,6 +787,7 @@ static void sc_wait_for_capture_helper(struct sc_mount_ns *group)
 	}
 	debug("helper process exited normally");
 	group->child = 0;
+	teardown_signals_for_helper();
```

Comment:
I misread where the teardown was placed; it is down in sc_wait_for_capture_helper() so my analysis was correct and the code is correct in that the parent needs the adjustment for the SIGPIPE. Sorry about misreading that.

Couple of things:
* I think that setup_signals_for_helper() need only be in the parent, no?
* The description focuses a lot on SIGCHLD when it should instead be talking more about SIGPIPE (why are we talking about restarting syscalls when it is simply that the child terminated from SIGALRM?)
* More code comments on why we need to alter SIGPIPE would be helpful
* Please investigate the read() returning 0.

Reply:
* setup_signals_for_helper(): I looked into this and played with it and I agree that we need the setup_signal_for_helper() only in the parent. I also tested this empirically by setting the signal helper only in the parent and sending a SIGPIPE to the helper. This does not change the exit code of the parent.
* use sigaction instead of signal: see above, AFAICT the usage of signal() in our case is ok
* read() returning 0: good point, I changed the code to to deal with that (and die() with a different error)

Reply:
we need to rework/simplify the description/commit message of of the PR before landing, that I think what jdstrand was referring to